packages/next/src/server/base-server.ts

```diff
@@ -2371,6 +2371,14 @@ export default abstract class Server<ServerOptions extends Options = Options> {
     ctx: RequestContext,
     err: Error | null
   ): Promise<ResponsePayload | null> {
+    // Short-circuit favicon.ico in development to avoid compiling 404 page when the app has no favicon.ico.
+    // Since favicon.ico is automatically requested by the browser.
+    if (this.renderOpts.dev && ctx.pathname === '/favicon.ico') {
```

Comment:
Maybe short-circuit in production is also a good idea. It wastes server to render full html 404 page that browser can't load as icon